logic/tool/TypeResolver.java

```diff
@@ -144,355 +83,174 @@ public Conjunction resolveLabels(Conjunction conjunction) {
-        visited.add(variable);
-        ensureResolvedConformToTheirSuper(above(variable), visited);
-        removeVarsViolatingSuper(variable);
-    }
-
-    private void removeVarsViolatingSuper(Variable subVar) {
-        TypeVariable supVar = above(subVar);
-        if (supVar == null) return;
-        if (supVar.reference().isLabel()) return;
-
-        Set<Label> supLabels = supVar.resolvedTypes();
-        Set<Label> subLabels = subVar.resolvedTypes();
-        if (supLabels.isEmpty()) return;
-
-        if (subLabels.isEmpty()) {
-            Set<Label> subResolvedOfSupLabels = supLabels.stream().flatMap(label -> getType(label).getSubtypes())
-                    .map(Type::getLabel).collect(Collectors.toSet());
-            subVar.addResolvedTypes(subResolvedOfSupLabels);
-            return;
-        }
+    public Conjunction resolveVariables(Conjunction conjunction) {
+        resolveLabels(conjunction);
+        TraversalConstructor traversalConstructor = new TraversalConstructor(conjunction, conceptMgr);
 
-        Set<Label> temp = new HashSet<>(subLabels);
-        for (Label label : temp) {
-            Type ResolvedType = getType(label);
-            while (ResolvedType != null && !supLabels.contains(ResolvedType.getLabel())) {
-                ResolvedType = ResolvedType.getSupertype();
-            }
-            if (ResolvedType == null) {
-                subVar.removeResolvedType(label);
-                if (subVar.resolvedTypes().isEmpty()) subVar.setSatisfiable(false);
-            }
-        }
-    }
+        Map<Reference, Set<Label>> resolvedLabels = runTraversalEngine(traversalConstructor);
+        if (resolvedLabels.isEmpty()) throw GraknException.of(SCHEMATICALLY_UNSATISFIABLE_CONJUNCTION, conjunction);
+        resolvedLabels.forEach((ref, labels) -> traversalConstructor.getVariable(ref).addResolvedTypes(labels));
+        findVariablesWithNoHints(conjunction);
 
-    private void addInferredSubLabels(TypeVariable variable, Set<Label> resolvedLabels) {
-        variable.addResolvedTypes(resolvedLabels);
+        return conjunction;
     }
 
-    private void addInferredIsaLabels(ThingVariable variable, Set<Label> resolveLabels) {
-        //TODO: use .getType(label) once ConceptManager can handle labels
-        resolveLabels.removeIf(label -> traversalEng.graph().schema().getType(label).isAbstract());
-        variable.addResolvedTypes(resolveLabels);
+    private void findVariablesWithNoHints(Conjunction conjunction) {
+        long numOfTypes = traversalEng.graph().schema().stats().thingTypeCount();
+        long numOfConcreteTypes = traversalEng.graph().schema().stats().concreteThingTypeCount();
+        iterate(conjunction.variables()).filter(variable -> (
+                variable.isType() && variable.resolvedTypes().size() == numOfTypes ||
+                        variable.isThing() && variable.resolvedTypes().size() == numOfConcreteTypes
+        )).forEachRemaining(Variable::clearResolvedTypes);
```

Comment:
Here we clear `resolvedTypes` for any variables that find every single type resolvable, and so the `resolvedTypes` bear no information. `Things` have had all abstract types removed, so we don't count those when considering the number of total types.